presto-spi/src/main/java/com/facebook/presto/spi/block/LazySliceArrayBlock.java

```diff
@@ -205,31 +212,39 @@ public String toString()
 
     private Block compactAndGet(List<Integer> positions, boolean copy)
     {
-        List<Integer> distinctPositions = positions.stream().distinct().collect(toList());
-        List<Integer> currentDictionaryIndexes = distinctPositions.stream().map(this::getPosition).collect(toList());
-        List<Integer> positionsToCopy = currentDictionaryIndexes.stream().distinct().collect(toList());
+        List<Integer> dictionaryIndexes = new ArrayList<>();
+        int[] newIds = new int[positions.size()];
+        boolean hasNull = false;
 
-        Slice[] newValues = new Slice[positionsToCopy.size()];
-        for (int i = 0; i < positionsToCopy.size(); i++) {
-            int position = positionsToCopy.get(i);
+        for (int i = 0; i < positions.size(); i++) {
+            int position = positions.get(i);
             if (isEntryNull(position)) {
-                newValues[i] = null;
+                hasNull = true;
+                newIds[i] = -1;
             }
             else {
-                Slice value = values[position];
-                if (copy) {
-                    newValues[i] = copyOf(value);
-                }
-                else {
-                    newValues[i] = value;
+                int oldIndex = ids[position];
+                if (!dictionaryIndexes.contains(oldIndex)) {
```

Comment:
this is a linear scan so you have an n^2 implementation.  Instead, I would keep a mapping from oldId -> newId in an array.